### PR TITLE
Add support for z13 processor

### PR DESCRIPTION
**Pull request by nealef**
The z13 supports some useful instructions that can be exploited by the JIT. Most notably the compare-and-branch instructions can reduce a two instruction sequence to one. The load-and-add command eliminates the more complex compare-and-swap sequence. The z13 also supports three operand instructions dst,src1,src2 which eliminate checks and loads for instructions like add and subtract.
The unaligned.cs test has been changed to cater for big-endian platforms.
For s390x only, the --trace option will result in a file per thread being created rather than the traces being written to stdout and sometimes intermixing incomplete trace entries.

### CODE REVIEW

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -462,8 +462,13 @@ typedef union {
 #define SPECIAL_STATIC_OFFSET_TYPE_THREAD 0
 #define SPECIAL_STATIC_OFFSET_TYPE_CONTEXT 1
 
+#if G_BYTE_ORDER != G_LITTLE_ENDIAN
+#define MAKE_SPECIAL_STATIC_OFFSET(index, offset, type) \
+	((SpecialStaticOffset) { .fields = { (type), (offset), (index) } }.raw)
```
Comment (alexrp) [inline mono/metadata/threads.c]: Wasn't this already fixed in #5398 ?

**Comment by nealef** [comment]
It was and I pulled from master and sent it to my remote so I thought it would disappear from the PR. How do I get rid of it?

@alexrp commented on this pull request.
________________________________

In mono/metadata/threads.c<https://github.com/mono/mono/pull/5420#discussion_r134529189>:

> @@ -462,8 ﮎ,13 @@ typedef union {
 #define SPECIAL_STATIC_OFFSET_TYPE_THREAD 0
 #define SPECIAL_STATIC_OFFSET_TYPE_CONTEXT 1

#if G_BYTE_ORDER != G_LITTLE_ENDIAN
#define MAKE_SPECIAL_STATIC_OFFSET(index, offset, type) \
        ((SpecialStaticOffset) { .fields = { (type), (offset), (index) } }.raw)


Wasn't this already fixed in #5398<https://github.com/mono/mono/pull/5398> ?

**Comment by akoeplinger** [comment]
@nealef it'll go away if you rebase (instead of merge) on our master branch.

**Comment by nealef** [comment]
I hate when I get into strange states with git. I rebased and skipped the threads.c change - 

$ git status
# On branch z13v5
# Your branch and 'z13/z13v5' have diverged,
# and have 27 and 36 different commit(s) each, respectively.
#
nothing added to commit 
[neale@aussie-1 mono]$ git push -u z13 
Password: 
To https://nealef@github.com/nealef/mono.git
 ! [rejected]        z13v5 -> z13v5 (non-fast-forward)
Branch master set up to track remote branch master from z13.
error: failed to push some refs to 'https://nealef@github.com/nealef/mono.git'
To prevent you from losing history, non-fast-forward updates were rejected
Merge the remote changes before pushing again.  See the 'Note about
fast-forwards' section of 'git push --help' for details.

**Comment by nealef** [comment]
How do I get past this?

**Comment by akoeplinger** [comment]
@nealef if your local branch z13v5 looks good then you can push with the `--force` flag to override the remote changes.

**Comment by nealef** [comment]
I'm not sure why mono/tests/Makefile.am appeared in my PR.

**Comment by akoeplinger** [comment]
@nealef it looks like something went wrong during your rebase. I pushed a cleaned up version to your branch.

**Comment by nealef** [comment]
Thanks, do I need to do anything?

File: mono/mini/unaligned.cs [hunk 2]
```diff
@@ -108,26 +133,51 @@ public static unsafe int test_0_ldind ()
 			ptr [i] = (byte)i;
 
 
-		if (Intrinsics.UnalignedLdInd2 (ptr + 0) != 0x0100)
-			return 1;
+		if (BitConverter.IsLittleEndian) {
 
-		if (Intrinsics.UnalignedLdInd2 (ptr + 1) != 0x0201)
-			return 2;
+			if (Intrinsics.UnalignedLdobj<short> (ptr + 0) != 0x0100)
```
Comment (akoeplinger) [inline mono/mini/unaligned.cs]: @nealef this changes here from `Intrinsics.UnalignedLdInd2` to `Intrinsics.UnalignedLdobj<short>`, I assume that's a copy-paste error?

Comment (nealef) [inline mono/mini/unaligned.cs]: Yes, I will fix and push.

**Comment by akoeplinger** [comment]
@nealef had to rebase your commits again since your new push only contained the unaligned.cs changes and removed the others :)

**Comment by nealef** [comment]
Sigh, one day I will get this right.

**Comment by akoeplinger** [comment]
don't worry, git can be intimidating sometimes 😄

Good practice is to always rebase your feature branch on top of master instead of merging master into your branch, then you shouldn't run into these issues, though

**Comment by nealef** [comment]
I will add that to my notes. I had been in the habit of git pull from master to pick up latest changes.

**Comment by nealef** [comment]
Thanks for the hand-holding folks.

**Pull request merged**